homeassistant/components/sensor/mysensors.py

```diff
@@ -11,146 +11,173 @@
-                    node[child_id][value_type] = \
-                        MySensorsNodeValue(
-                            nid, child_id, name, value_type, is_metric, const)
-                    new_devices.append(node[child_id][value_type])
-                else:
-                    node[child_id][value_type].update_sensor(
-                        value, sensor.battery_level)
-
-        if new_devices:
-            _LOGGER.info("adding new devices: %s", new_devices)
-            add_devices(new_devices)
-
-    port = config.get(CONF_PORT)
-    if port is None:
-        _LOGGER.error("Missing required key 'port'")
-        return False
+DEPENDENCIES = []
 
-    persistence = config.get(CONF_PERSISTENCE, True)
-    persistence_file = config.get(CONF_PERSISTENCE_FILE,
-                                  hass.config.path('mysensors.pickle'))
-    version = config.get(CONF_VERSION, '1.4')
+ADD_DEVICES = None
+S_TYPES = None
+V_TYPES = None
 
-    gateway = mysensors.SerialGateway(port, sensor_update,
-                                      persistence=persistence,
-                                      persistence_file=persistence_file,
-                                      protocol_version=version)
-    gateway.metric = is_metric
-    gateway.debug = config.get(CONF_DEBUG, False)
-    gateway.start()
 
-    if persistence:
-        for nid in gateway.sensors:
-            sensor_update('sensor_update', nid)
+@mysensors.mysensors_update
+def sensor_update(gateway, port, devices, nid):
+    """Internal callback for sensor updates."""
+    return (S_TYPES, V_TYPES, MySensorsSensor, ADD_DEVICES)
 
-    hass.bus.listen_once(EVENT_HOMEASSISTANT_STOP,
-                         lambda event: gateway.stop())
 
+def setup_platform(hass, config, add_devices, discovery_info=None):
+    """Setup the mysensors platform for sensors."""
+    # Define the S_TYPES and V_TYPES that the platform should handle as states.
+    global ADD_DEVICES
+    ADD_DEVICES = add_devices
+    global S_TYPES
+    S_TYPES = [
+        mysensors.CONST.Presentation.S_TEMP,
+        mysensors.CONST.Presentation.S_HUM,
+        mysensors.CONST.Presentation.S_BARO,
+        mysensors.CONST.Presentation.S_WIND,
+        mysensors.CONST.Presentation.S_RAIN,
+        mysensors.CONST.Presentation.S_UV,
+        mysensors.CONST.Presentation.S_WEIGHT,
+        mysensors.CONST.Presentation.S_POWER,
+        mysensors.CONST.Presentation.S_DISTANCE,
+        mysensors.CONST.Presentation.S_LIGHT_LEVEL,
+        mysensors.CONST.Presentation.S_IR,
+        mysensors.CONST.Presentation.S_WATER,
+        mysensors.CONST.Presentation.S_AIR_QUALITY,
+        mysensors.CONST.Presentation.S_CUSTOM,
+        mysensors.CONST.Presentation.S_DUST,
+        mysensors.CONST.Presentation.S_SCENE_CONTROLLER,
+    ]
+    not_v_types = [
+        mysensors.CONST.SetReq.V_ARMED,
+        mysensors.CONST.SetReq.V_LIGHT,
+        mysensors.CONST.SetReq.V_LOCK_STATUS,
+    ]
+    if float(mysensors.VERSION) >= 1.5:
```

Comment:
If you're using it as a float here, might be a good idea to just store the version as a float to begin with?